### ISSUE TITLE
False positive UnsafeCallOnNullableType when using e.g. exitProcess()

### ISSUE THREAD

**Issue by sschuberth** [arturbosch/detekt#202]
Code like

        if (projectPaths == null) {
            exitProcess(1)
        }

        // ...

        projectPaths!!.forEach { projectPath ->
        }

triggers `UnsafeCallOnNullableType` although `projectPaths` cannot be `null` at that point as `exitProcess()` never returns (it is marked to return `Nothing`).

**Comment by arturbosch** [comment]
I would just say it is planned. Ah full data flow analysis is impossible as you can extract sub routines which do the null check or some other checks -> performance flow downs.

It's a turned off rule to let you check all the locations if you want to refactor something. 

Does intellij flag the !! As unnecessary?

**Comment by sschuberth** [comment]
> Does intellij flag the !! As unnecessary?

No, it doesn't, it actually demands it (which I'd also consider a bug, in Kotlin itself). But you're right, this is probably too hard to fix on the detekt level.

**Comment by MyDogTom** [comment]
@sschuberth Make sure your `exitProcess` has type `Nothing` and not `Unit`. Here is an example to show that Kotlin understand it correctly
```kotlin
    fun neverReturn(): Nothing {
        while (true) { }
    }

    fun testNever(testVar: List<String>?) {
        if (testVar == null) {
            neverReturn()
        }
        testVar.forEach { it.length }
    }
```

**Comment by sschuberth** [comment]
@MyDogTom As written in my initial post it *does* return `Nothing`, and it's not "my" `exitProcess()` but the one from `kotlin.system`.
